src/S3Url.php

```diff
@@ -119,18 +118,38 @@ public function getImageStyleUrl($styleName) {
    * @return static
    *   An S3Url.
    */
-  public static function factory($url, StreamWrapperConfiguration $config = null) {
-    !$config ? $bucket = null : $bucket = $config->getBucket();
-
-    $defaults = array('scheme' => 's3', 'host' => $bucket, 'path' => null, 'port' => null, 'query' => null,
-      'user' => null, 'pass' => null, 'fragment' => null);
-
-    if (false === ($parts = parse_url($url))) {
+  public static function factory($url, StreamWrapperConfiguration $config = NULL) {
+    $defaults = array(
+      'scheme' => 's3',
+      'host' => $config ? $config->getBucket() : NULL,
+      'path' => NULL,
+      'port' => NULL,
+      'query' => NULL,
+      'user' => NULL,
+      'pass' => NULL,
+      'fragment' => NULL,
+    );
+
+    $parts = parse_url($url);
+    if ($parts === FALSE) {
       throw new \InvalidArgumentException('Was unable to parse malformed url: ' . $url);
     }
-
     $parts += $defaults;
 
+    // Image styles are constructed as
```

Comment:
> Conceptually it seems like figuring out $parts['path'] is something we would want to reuse somewhere else.

Sure I can do that. I suppose it would also be a static method, as the factory is also.

> For that matter, would it be clearer to have an S3ImageStyleUrl class that overrides factory()?

I would say no, not unless we have another class that would also extend S3Url. At this point it would be premature abstraction.